model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala

```diff
@@ -373,6 +371,56 @@ object ArrowInliner extends Logging {
     } yield values -> arrows
   }
 
+  //
+  private def collectionsToCanons(
+    tree: RawTag.Tree,
+    streamArgs: Map[String, VarModel]
+  ): RawTag.Tree = {
+    // collect arguments with stream type
+    // to exclude it from resolving and rename it with a higher-level stream that passed by argument
+    val streamsToRename = streamArgs.view.mapValues(_.name).toMap
+
+    if (streamsToRename.isEmpty) tree
+    else
+      tree
+        .map(_.mapValues(_.map {
+          // if an argument is a BoxType (Array or Option), but we pass a stream,
+          // change a type as stream to not miss `$` sign in air
+          case v @ VarRaw(name, baseType: CollectionType) if streamsToRename.contains(name) =>
+            v.copy(baseType = CanonStreamType(baseType.element))
+          case v => v
+        }))
+        .renameExports(streamsToRename)
+  }
+
+  // Canonicalize streams that passed to function as immutable collection variables
+  private def canonStreamVariables[S: Mangler](
+    args: ArgsCall
+  ): State[S, (List[CanonicalizeModel#Tree], Map[String, String])] = {
+    val streamToImmutableArgs = args.streamToImmutableArgs
+    for {
+      newStreamsCanon <- streamToImmutableArgs.values.toList.collect {
+        case vm @ VarModel(name, st: StreamType, _) =>
+          (vm, name, st)
+      }.traverse { case (vm, name, StreamType(t)) =>
+        Mangler[S].findAndForbidName(name + "_canon").map { n =>
+          val canonVM = VarModel(n, CanonStreamType(t))
+          (
+            name,
+            canonVM.name,
+            CanonicalizeModel(vm, CallModel.Export(canonVM.name, canonVM.`type`)).leaf
+          )
+        }
+      }
+      canons = newStreamsCanon.map(_._3)
+      streamsToImmutableRenames = args.streamToImmutableArgsRenames
+      streamsToImmutableA = newStreamsCanon.map(kv => kv._1 -> kv._2).toMap
+      renamedCanonStreams = streamsToImmutableRenames.map(kv =>
+        kv._1 -> streamsToImmutableA.getOrElse(kv._2, kv._2)
+      )
```

Comment:
This code is really hard to understand. If I got this right, you try to unite `arg name -> passed stream name` and `passed stream name -> canon name` to `arg name -> canon name`. I think you can retrieve `arg name` in the `traverse` above and just return it instead of `name`. This way this map is constructed automatically in `traverse`